dojo/endpoint/utils.py

```diff
@@ -127,89 +138,101 @@ def clean_hosts_run(apps, change):
-                                    'fragment in host ({}). {}'.format(endpoint.pk, endpoint.fragment, parts.fragment,
-                                                                       error_suffix))
-                                broken_endpoints.append(endpoint.pk)
+                                message = 'has defined fragment ({}) and it is not the same as fragment in host ' \
+                                          '({})'.format(endpoint.fragment, parts.fragment)
+                                html_log.append({**endpoint_html_log, **{'message': message}})
+                                logger.error('Endpoint (id={}) {}. {}'.format(endpoint.pk, message, error_suffix))
+                                broken_endpoints.add(endpoint.pk)
                             else:
                                 if change:
                                     endpoint.fragment = parts.fragment
 
-                        if change:
+                        if change and (endpoint.pk not in broken_endpoints):  # do not save broken endpoints
                             endpoint.save()
 
                     except ValidationError:
-                        logger.error('Endpoint (id={}) "{}" use invalid format of host. {}'.format(endpoint.pk,
-                            endpoint.host, error_suffix))
-                        broken_endpoints.append(endpoint.pk)
-
-    if broken_endpoints != []:
-        raise FieldError('It is not possible to migrate database because there is/are {} broken endpoint(s). '
-                         'Please check logs.'.format(len(broken_endpoints)))
-
-    if change:
-        to_be_deleted = set()
-        for product in chain(Product_model.objects.all().distinct(), [None]):
-            for endpoint in Endpoint_model.objects.filter(product=product).distinct():
-                if endpoint.id not in to_be_deleted:
-
-                    ep = endpoint_filter(
-                        protocol=endpoint.protocol,
-                        userinfo=endpoint.userinfo,
-                        host=endpoint.host,
-                        port=endpoint.port,
-                        path=endpoint.path,
-                        query=endpoint.query,
-                        fragment=endpoint.fragment,
-                        product_id=product.pk if product else None
-                    ).order_by('id')
-
-                    if ep.count() > 1:
-                        ep_ids = [x.id for x in ep]
-                        logger.info("Merging Endpoints {} into '{}'".format(
+                        message = '"{}" uses invalid format of host'.format(endpoint.host)
+                        html_log.append({**endpoint_html_log, **{'message': message}})
+                        logger.error('Endpoint (id={}) {}. {}'.format(endpoint.pk, message, error_suffix))
+                        broken_endpoints.add(endpoint.pk)
+
+    if broken_endpoints != set():
```

Comment:
is this reliable? I think it could just be `if broken_endpoints:`

Reply:
Python says, `if broken_endpoints:` should be enought.
```python
>>> bool(set())
False
```

Reply:
I agree it works but it's not the best code to read. :/ should we use the form `len(broken_endpoints)` ? better for readability.

Reply:
```
>>> empty_valentijn = set()
>>> if empty_valentijn:
...     print('not empty')
... else:
...     print('empty')
...
empty
```

```
>>> valentijn = set()
>>> valentijn.add('test')
>>> if valentijn:
...     print('not empty')
...
not empty
```